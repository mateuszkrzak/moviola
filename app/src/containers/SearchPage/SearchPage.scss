@import "../../common/styles/variables.scss";

.find-button {
  padding: 10px 50px;
  background-color: $pink;
  margin-left: auto;
}

.actions {
  margin-top: 10px;
  display: flex;
  align-items: center;
}
