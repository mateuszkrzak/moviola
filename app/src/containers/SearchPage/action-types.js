export const FETCH_MOVIES_REQUEST = 'FETCH_MOVIES_REQUEST';
export const FETCH_MOVIES_FAILURE = 'FETCH_MOVIES_FAILURE';
export const FETCH_MOVIES_SUCCESS = 'FETCH_MOVIES_SUCCESS';
export const SET_MOVIES_QUERY = 'SET_MOVIES_QUERY';
export const SET_MOVIES_SORT_BY = 'SET_MOVIES_SORT_BY';
export const SET_MOVIES_SEARCH_BY = 'SET_MOVIES_SEARCH_BY';