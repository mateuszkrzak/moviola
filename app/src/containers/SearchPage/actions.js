// @flow

import moviesService from '../../services/MoviesService';
import type { Movie } from './reducer';
import { 
  FETCH_MOVIES_FAILURE, 
  FETCH_MOVIES_REQUEST, 
  FETCH_MOVIES_SUCCESS, 
  SET_MOVIES_QUERY, 
  SET_MOVIES_SORT_BY, 
  SET_MOVIES_SEARCH_BY,
} from './action-types';

type FetchMoviesRequestAction = {| 
  +type: typeof FETCH_MOVIES_REQUEST 
|};
type FetchMoviesFailureAction = {| 
  +type: typeof FETCH_MOVIES_FAILURE, 
  +payload: string 
|};
type FetchMoviesSuccessAction = {| 
  +type: typeof FETCH_MOVIES_SUCCESS, 
  +payload: { 
    +movies: Array<Movie>,
    +moviesCount: number
  } 
|};
export type SetMoviesQueryAction = {|
  +type: typeof SET_MOVIES_QUERY, 
  +payload: string 
|};
export type SetMoviesSortByAction = {|
  +type: typeof SET_MOVIES_SORT_BY, 
  +payload: string 
|};
export type SetMoviesSearchByAction = {|
  +type: typeof SET_MOVIES_SEARCH_BY, 
  +payload: string 
|};

function getMoviesRequest(): FetchMoviesRequestAction {
  return {
    type: FETCH_MOVIES_FAILURE,
  };
}

function getMoviesFailure(error: string): FetchMoviesFailureAction {
  return {
    type: FETCH_MOVIES_FAILURE,
    payload: error,
  };
}



function getMoviesSuccess({ data, total }: *): FetchMoviesSuccessAction {
  return {
    type: FETCH_MOVIES_SUCCESS,
    payload: { 
      movies: data, 
      moviesCount: total,
    },
  };
}

function getMoviesAsync() {
  return async (dispatch, getState) => {
    dispatch(getMoviesRequest());
    try {
      const { query, searchBy, sortBy } = getState().search;
      const response = await moviesService.getMovies({ searchBy, sortBy, search: query });
      dispatch(getMoviesSuccess(response.data));
    } catch (error) {
      dispatch(getMoviesFailure(error.message));
    }
  };
}

function setMoviesQuery(query: string): SetMoviesQueryAction {
  return {
    type: SET_MOVIES_QUERY,
    payload: query,
  };
}
function setMoviesSortBy(sortBy: string): SetMoviesSortByAction {
  return {
    type: SET_MOVIES_SORT_BY,
    payload: sortBy,
  };
}
function setMoviesSearchBy(searchBy: string): SetMoviesSearchByAction {
  return {
    type: SET_MOVIES_SEARCH_BY,
    payload: searchBy,
  };
}

export { getMoviesAsync, setMoviesQuery, setMoviesSortBy, setMoviesSearchBy };
