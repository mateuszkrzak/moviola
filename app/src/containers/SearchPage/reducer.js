import { 
  FETCH_MOVIES_REQUEST, 
  FETCH_MOVIES_FAILURE, 
  FETCH_MOVIES_SUCCESS,
  SET_MOVIES_QUERY,
  SET_MOVIES_SORT_BY,
  SET_MOVIES_SEARCH_BY
} from './action-types';
import createReducer from '../../redux/create-reducer';

const initialState = {
  movies: [],
  error: '',
  isLoading: false,
  moviesCount: 0,
  query: '',
  sortBy: '',
  searchBy: 'title'
};

const fetchMoviesRequest = state => ({ ...state, isLoading: true });
const fetchMoviesFailure = (state, payload) => ({
  ...state,
  isLoading: false,
  error: payload,
});
const fetchMoviesSuccess = (state, payload) => ({
  ...state,
  movies: [...payload.movies],
  moviesCount: payload.moviesCount,
  isLoading: false,
  error: '',
});

const setMoviesQuery = (state, payload) => ({
  ...state,
  query: payload
})
const setMoviesSortBy = (state, payload) => ({
  ...state,
  sortBy: payload
})
const setMoviesSearchBy = (state, payload) => ({
  ...state,
  searchBy: payload
})

export default createReducer(initialState, {
  [FETCH_MOVIES_REQUEST]: fetchMoviesRequest,
  [FETCH_MOVIES_FAILURE]: fetchMoviesFailure,
  [FETCH_MOVIES_SUCCESS]: fetchMoviesSuccess,
  [SET_MOVIES_QUERY]: setMoviesQuery,
  [SET_MOVIES_SORT_BY]: setMoviesSortBy,
  [SET_MOVIES_SEARCH_BY]: setMoviesSearchBy
});
