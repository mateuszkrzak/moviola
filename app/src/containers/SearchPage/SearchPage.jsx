import React from 'react';
import { connect } from 'react-redux';
import * as qs from 'qs';

import './SearchPage.scss';
import Logo from '../../common/components/Logo/Logo';
import SearchBar from './SearchBar/SearchBar';
import SearchFilters from './SearchFilters/SearchFilters';
import MovieList from '../../common/components/MovieList/MovieList';
import MovieListOptions from './MovieListOptions/MovieListOptions';

import { getMoviesAsync, setMoviesQuery, setMoviesSearchBy, setMoviesSortBy } from './actions';

export class SearchPage extends React.Component {
  componentDidMount() {
    const query = this.props.location.pathname.replace('/search/', '');

    if (query !== '/' && this.props.query === '') {    
      const searchValues = qs.parse(query);        
      this.props.setMoviesQuery(searchValues.search);
      this.props.setMoviesSearchBy(searchValues.searchBy);
      this.props.setMoviesSortBy(searchValues.sortBy);
      this.getMovies();
      }
  }

  getMovies() {
    this.props.getMoviesAsync();

    if (this.props.search !== '') {      
      const suffix = this.props.match.params.query ? '' : 'search/';
      const query = suffix + qs.stringify({
        search: this.props.search,
        searchBy: this.props.searchBy,
        sortBy: this.props.sortBy,
      });

      this.props.history.push(query);
    }
  }

  searchValueChangeHandler = (search) => {
    this.props.setMoviesQuery(search);
  };

  searchFiltersChangeHandler = (searchBy) => {
    this.props.setMoviesSearchBy(searchBy);
  };

  searchActionHandler = () => {
    this.getMovies();
  };

  sortChangeHandler = (sortBy) => {
    this.props.setMoviesSortBy(sortBy);
    this.getMovies();
  };

  render() {
    return (
      <div className="page-wrapper">
        <main className="page-header">
          <Logo />
          <SearchBar value={this.props.search} onChange={this.searchValueChangeHandler} onSubmit={this.searchActionHandler} />
          <div styleName="actions">
            <SearchFilters value={this.props.searchBy} onChange={this.searchFiltersChangeHandler} />
            <button
              onClick={this.searchActionHandler}
              data-qa="search-button"
              className="standard-button"
              styleName="find-button"
            >
              Find
            </button>
          </div>
        </main>
        <MovieListOptions onChange={this.sortChangeHandler} moviesCount={this.props.moviesCount} />
        <MovieList movies={this.props.movies} />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  movies: state.search.movies,
  moviesCount: state.search.moviesCount,
  search: state.search.query,
  searchBy: state.search.searchBy,
  sortBy: state.search.sortBy,
});

export default connect(mapStateToProps, { getMoviesAsync, setMoviesQuery, setMoviesSearchBy, setMoviesSortBy })(SearchPage);
