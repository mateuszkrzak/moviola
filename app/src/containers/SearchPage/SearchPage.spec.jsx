import React from 'react';
import { shallow } from 'enzyme';
import { SearchPage } from './SearchPage';

describe('SearchPage component ', () => {
  test('renders SearchPage', () => {
    const component = shallow(<SearchPage />);

    expect(component).toMatchSnapshot();
  });
});
