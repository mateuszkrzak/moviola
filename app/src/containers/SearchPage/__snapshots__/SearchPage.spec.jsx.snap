// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage component renders SearchPage 1`] = `
<div
  className="page-wrapper"
>
  <main
    className="page-header"
  >
    <span
      styleName="title"
    >
      netflixroulette
    </span>
    <h2
      styleName="title"
    >
      Find your movie
    </h2>
    <input
      data-qa="search-bar"
      placeholder="Type something"
      styleName="search-box"
    />
    <div
      styleName="actions"
    >
      <div
        styleName="wrapper"
      >
        <span
          styleName="title"
        >
          Search by
        </span>
        <button
          className="standard-button"
        >
          Title
        </button>
        <button
          className="standard-button"
          styleName="active"
        >
          Genre
        </button>
      </div>
      <button
        className="standard-button"
        data-qa="search-button"
        styleName="find-button"
      >
        Find
      </button>
    </div>
  </main>
  <div
    styleName="wrapper"
  >
    <span
      styleName="title"
    >
      10
       movies found
    </span>
    <div
      styleName="wrapper"
    >
      <span
        styleName="title"
      >
        Sort by
      </span>
      <span
        styleName="sort-type"
      >
        release date
      </span>
      <span
        styleName="sort-type active"
      >
        rating
      </span>
    </div>
  </div>
  <div
    data-qa="movie-list"
    styleName="list"
  >
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Aquaman"
        src="https://image.tmdb.org/t/p/w500/sshNRCbtU2ON8SiWLycQzr05doX.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Aquaman
           
        </span>
        <span
          styleName="title"
        >
          2018
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Fantasy & Science Fiction
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Guardians of the Galaxy Vol. 3"
        src="https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Guardians of the Galaxy Vol. 3
           
        </span>
        <span
          styleName="title"
        >
          2020
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Adventure & Science Fiction
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Transformers 7"
        src="https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Transformers 7
           
        </span>
        <span
          styleName="title"
        >
          2019
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Science Fiction & Action & Adventure
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Untitled Avengers"
        src="https://image.tmdb.org/t/p/w500/aMETsaNNcDc6g5ZatQtVbySnSaA.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Untitled Avengers
           
        </span>
        <span
          styleName="title"
        >
          2019
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Adventure & Science Fiction
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Shazam!"
        src="https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Shazam!
           
        </span>
        <span
          styleName="title"
        >
          2019
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Adventure & Fantasy
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of X-Men: Dark Phoenix"
        src="https://image.tmdb.org/t/p/w500/6qmsupE0opYPIaBGe7T5D2FBzLs.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          X-Men: Dark Phoenix
           
        </span>
        <span
          styleName="title"
        >
          2019
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Science Fiction
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Journey 3: From the Earth to the Moon"
        src="https://image.tmdb.org/t/p/w500/98tbNloMBztAVnWpAznKKVUdi2O.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Journey 3: From the Earth to the Moon
           
        </span>
        <span
          styleName="title"
        >
          2018
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Fantasy
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Mission: Impossible - Fallout"
        src="https://image.tmdb.org/t/p/w500/A2W75CVz9l8gEeWhQiiftlOCkgi.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Mission: Impossible - Fallout
           
        </span>
        <span
          styleName="title"
        >
          2018
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Adventure & Action & Thriller
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Jurassic World: Fallen Kingdom"
        src="https://image.tmdb.org/t/p/w500/ln6d5Okr6VK5vfQVobJTiYxeD0l.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Jurassic World: Fallen Kingdom
           
        </span>
        <span
          styleName="title"
        >
          2018
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Adventure & Drama & Thriller
         
      </span>
    </div>
    <div
      styleName="wrapper"
    >
      <img
        alt="Poster of Solo: A Star Wars Story"
        src="https://image.tmdb.org/t/p/w500/uJ6OnE3CzGWq6buLINAbdBqa0gV.jpg"
        styleName="poster"
      />
      <div
        styleName="row"
      >
        <span
          styleName="title"
        >
          Solo: A Star Wars Story
           
        </span>
        <span
          styleName="title"
        >
          2018
           
        </span>
      </div>
      <span
        styleName="genre"
      >
         
        Action & Adventure & Family
         
      </span>
    </div>
  </div>
</div>
`;
